solidity_contracts/src/CairoPrecompiles/MulticallCairoCounterCaller.sol

```diff
@@ -0,0 +1,209 @@
+    {
+        bytes4 selector = bytes4(keccak256("call_contract(bytes4,uint256,uint256,uint256[])"));
+        bytes memory callData =
+            abi.encodePacked(selector, bytes4(uint32(1)), abi.encode(contractAddress, functionSelector, data));
+
+        (bool success, bytes memory result) = BATCH_CAIRO_PRECOMPILE_ADDRESS.call(callData);
+        require(success, string(abi.encodePacked("CairoLib: call_contract failed with: ", result)));
+
+        return result;
+    }
+
+    function callCairo(uint256 contractAddress, uint256 functionSelector) internal returns (bytes memory) {
+        uint256[] memory data = new uint256[](0);
+        return callCairo(contractAddress, functionSelector, data);
+    }
+
+    function callCairo(uint256 contractAddress, string memory functionName, uint256[] memory data)
+        internal
+        returns (bytes memory)
+    {
+        uint256 functionSelector = uint256(keccak256(bytes(functionName))) % 2 ** 250;
+        return callCairo(contractAddress, functionSelector, data);
+    }
+
+    function callCairo(uint256 contractAddress, string memory functionName) internal returns (bytes memory) {
+        uint256[] memory data = new uint256[](0);
+        return callCairo(contractAddress, functionName, data);
+    }
+
+    function delegatecallCairo(uint256 contractAddress, uint256 functionSelector, uint256[] memory data)
+        internal
+        returns (bytes memory)
+    {
+        bytes4 selector = bytes4(keccak256("call_contract(bytes4,uint256,uint256,uint256[])"));
+        bytes memory callData =
+            abi.encodePacked(selector, bytes4(uint32(1)), abi.encode(contractAddress, functionSelector, data));
+
+        bool success;
+        bytes memory result;
+
+        (success, result) = BATCH_CAIRO_PRECOMPILE_ADDRESS.delegatecall(callData);
+
+        require(success, string(abi.encodePacked("CairoLib: call_contract failed with: ", result)));
+
+        return result;
+    }
+
+    function delegatecallCairo(uint256 contractAddress, uint256 functionSelector) internal returns (bytes memory) {
+        uint256[] memory data = new uint256[](0);
+        return delegatecallCairo(contractAddress, functionSelector, data);
+    }
+
+    function delegatecallCairo(uint256 contractAddress, string memory functionName, uint256[] memory data)
+        internal
+        returns (bytes memory)
+    {
+        uint256 functionSelector = uint256(keccak256(bytes(functionName))) % 2 ** 250;
+        return delegatecallCairo(contractAddress, functionSelector, data);
+    }
+
+    function delegatecallCairo(uint256 contractAddress, string memory functionName) internal returns (bytes memory) {
+        uint256[] memory data = new uint256[](0);
+        uint256 functionSelector = uint256(keccak256(bytes(functionName))) % 2 ** 250;
+        return delegatecallCairo(contractAddress, functionSelector, data);
+    }
+
+    function callcodeCairo(uint256 contractAddress, uint256 functionSelector, uint256[] memory data)
```

Comment:
Should callcode be authorized ?
It seems not necessary as it is advised to not use it, I suppose it is only for testing ?

Reply:
Yes, this entire file is just for testing purposes, as indicated in line 105